smithy-aws-traits/src/main/java/software/amazon/smithy/aws/traits/HttpChecksumTraitValidator.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.aws.traits;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import software.amazon.smithy.aws.traits.auth.SigV4Trait;
+import software.amazon.smithy.aws.traits.protocols.AwsProtocolTrait;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.knowledge.ServiceIndex;
+import software.amazon.smithy.model.knowledge.TopDownIndex;
+import software.amazon.smithy.model.shapes.OperationShape;
+import software.amazon.smithy.model.shapes.ServiceShape;
+import software.amazon.smithy.model.shapes.ShapeId;
+import software.amazon.smithy.model.traits.HttpChecksumProperties.Location;
+import software.amazon.smithy.model.traits.HttpChecksumTrait;
+import software.amazon.smithy.model.traits.OptionalAuthTrait;
+import software.amazon.smithy.model.traits.Trait;
+import software.amazon.smithy.model.validation.AbstractValidator;
+import software.amazon.smithy.model.validation.ValidationEvent;
+import software.amazon.smithy.utils.SmithyInternalApi;
+
+/**
+ * Validates checksum location modeling specific to AWS usage. For response property within httpChecksum trait,
+ * only "header" is a valid checksum location. If service, operation uses sigv4 authentication scheme, the
+ * request property within httpChecksum trait must include "header" as supported checksum location.
+ */
+@SmithyInternalApi
+public class HttpChecksumTraitValidator extends AbstractValidator {
+
+    @Override
+    public List<ValidationEvent> validate(Model model) {
+        List<ValidationEvent> events = new ArrayList<>();
+        ServiceIndex serviceIndex = ServiceIndex.of(model);
+
+        model.shapes(ServiceShape.class).filter(this::isTargetProtocol).forEach(service -> {
```

Comment:
Nit: We usually avoid chaining operations for simple cases like this as it's easier to read in the end.

```
TopDownIndex topDownIndex = TopDownIndex.of(model);
for (ServiceShape service : model.shapes(ServiceShape.class)) {
    if (!isTargetProtocol(service)) {
        continue;
    }

    for (OperationShape operation : topDownIndex.getContainedOperations(service)) {
        ...
    }
}
```